### PR TITLE
Fix for generating imports for MAIN_THREAD_EM_ASM

### PR DESCRIPTION
**Pull request by jgravelle-google**
Missed the part where we generate imports

### CODE REVIEW

File: emscripten.py [hunk 1]
```diff
@@ -2537,9 +2537,10 @@ def create_sending_wasm(invoke_funcs, forwarded_json, metadata):
   if shared.Settings.SAFE_HEAP:
     basic_funcs += ['segfault', 'alignfault']
 
-  em_asm_sigs = [sigs for _, sigs, _ in metadata['asmConsts'].values()]
+  em_asm_sigs = [zip(sigs, call_types) for _, sigs, call_types in metadata['asmConsts'].values()]
```
Comment (dschuff) [inline emscripten.py]: is the call_types part of this tuple actually used?

Comment (jgravelle-google) [reply emscripten.py]: Yeah, on the other changed line
This change turns em_asm_sigs from a list of signatures into a list of signature+call_type pairs, and we use both parts of the pair to generate the correct name

Comment (dschuff) [reply emscripten.py]: oh, right... i was just confused by the flattening step in between.

**Comment by hostilefork** [comment]
I'd like to test this, but when I did `emsdk install latest-upstream` and `emsdk activate latest-upstream` I did not notice any difference and MAIN_THREAD_EM_ASM still did not work.  I gather that did not pick up the `incoming` changes yet (?)

I tried [the suggestion here](https://brionv.com/log/2019/03/03/simd-in-webassembly-tales-from-the-bleeding-edge/) to explicitly ask for incoming:

    ./emsdk install emscripten-incoming-64bit
    ./emsdk activate emscripten-incoming-64bit
    ./emsdk install upstream-clang-master-64bit
    ./emsdk activate upstream-clang-master-64bit
    ./emsdk install binaryen-master-64bit
    ./emsdk activate binaryen-master-64bit

But the Python gets an error:

    Traceback (most recent call last):
      File "/home/hostilefork/emsdk/emscripten/incoming/emcc.py", line 41, in <module>
        import emscripten
      File "/home/hostilefork/emsdk/emscripten/incoming/emscripten.py", line 25, in <module>
        from tools import shared
      File "/home/hostilefork/emsdk/emscripten/incoming/tools/shared.py", line 760, in <module>
        CLANG_CC = os.path.expanduser(build_clang_tool_path(exe_suffix('clang')))
      File "/home/hostilefork/emsdk/emscripten/incoming/tools/shared.py", line 637, in build_clang_tool_path
        return os.path.join(LLVM_ROOT, tool)
      File "/usr/lib/python2.7/posixpath.py", line 70, in join
        elif path == '' or path.endswith('/'):
    AttributeError: 'NoneType' object has no attribute 'endswith'

My questions would be:

* Am I correct in saying that `latest-upstream` will not have this change?
* If so, what is the process by which this change would become available in `latest-upstream`?  (or how can I get past this `emscripten-incoming-64bit` problem--is it a bug, should I report it?)
* If I'm incorrect, should I look further into why MAIN_THREAD_EM_ASM is still not working?

**Comment by sbc100** [comment]
`emsdk install latest-upstream` should work once we publish a new emscripten release.   In the mean time you can try `tot-upstream` to get the very latest build.

In terms of the python error you are seeing, that is because `./emsdk activate binaryen-master-64bit` leaves the emscripten config without any LLVM_ROOT.  I've created a fix for that: https://github.com/emscripten-core/emscripten/pull/9606

**Comment by hostilefork** [comment]
Thanks, `tot-upstream` worked *(though it now needs one to say `./emsdk update-tags` instead of `./emsdk update`; it seemed to not work if this is not done even from a fresh `git clone`)*

My main thread call is asserting, however, at this line:

https://github.com/emscripten-core/emscripten/blob/3a142bf374063597fc8b3fa51f7718eccd41760d/src/library_pthread.js#L1253

The `index` is negative (-7 in my case), indicating an EM_ASM and not a proxied JS func.  The numCallArgs claims to be 3.  However the func.length reports 1.

Due to some complexities my building and testing is not as easy as it should be at the moment :-/ but I believe the routine being called in question is an MAIN_THREAD_EM_ASM_INT() with one argument ($0) and one return since it's the _INT() form.  I can file a more formal minimal case, but wanted to just mention this in case @jgravelle-google has any ideas offhand of what might be going wrong.  (Not lazy, just busy.  :-P)

**Comment by kripken** [comment]
@hostilefork a fix just landed for related things, can you try with latest `tot-upstream`?

**Pull request merged**